lib/graphql/schema/input_object.rb

```diff
@@ -184,21 +184,25 @@ def validate_non_null_input(input, ctx)
             end
           end
 
-          visible_arguments_map = warden.arguments(self).reduce({}) { |m, f| m[f.name] = f; m}
-
-          # Items in the input that are unexpected
-          input.each do |name, value|
-            if visible_arguments_map[name].nil?
-              result.add_problem("Field is not defined on #{self.graphql_name}", [name])
+          # Inject missing required arguments
+          required_inputs = self.arguments.reduce({}) do |m, (argument_name, argument)|
+            if input[argument_name].nil? && argument.type.non_null? && warden.get_argument(self, argument_name)
+              m[argument_name] = nil
             end
+
+            m
           end
 
-          # Items in the input that are expected, but have invalid values
-          visible_arguments_map.map do |name, argument|
-            argument_result = argument.type.validate_input(input[name], ctx)
-            if !argument_result.valid?
-              result.merge_result!(name, argument_result)
+          input.merge(required_inputs).each do |argument_name, value|
```

Comment:
if we merged this the other way around we could build `required_inputs` much more simply?

Reply:
I think the solution here using select looks something like:

```ruby
required_inputs = self.arguments.select do |argument_name, argument|
  !input.key?(argument_name) && argument.type.non_null? && warden.get_argument(self, argument_name)
end.map { |argument| nil }
```

I don't think this is too much better than the reduce, but can change if necessary.